File: app/platform/fabric/e2e-test/chaincodes/shimApiDriver/node/shimApiDriver.js
/*
# Copyright IBM Corp. All Rights Reserved.
#
# SPDX-License-Identifier: Apache-2.0
*/

// This chaincode demonstrates the following ChaincodeStubInterface APIs
// More details can be obtained from the following link:
// https://github.com/hyperledger/fabric-chaincode-node/blob/master/src/lib/stub.js
//
// getArgs -- arguments intended for the chaincode Init and Invoke
//            as an array of byte arrays
// getStringArgs -- arguments intended for the chaincode Init and
//                 Invoke as a string array
// getFunctionAndParameters -- returns the first argument as the function
//                             name and the rest of the arguments as parameters
//                             in a string array
// getTxID -- the Transaction ID of the transaction proposal
// invokeChaincode -- call the specified chaincode `Invoke` using the
//                     same transaction context
// getCreator -- returns `SignatureHeader.Creator` (e.g. an identity)
//               of the `SignedProposal`
// getTransient -- It is a map that contains data (e.g. cryptographic material)
//                that might be used to implement some form of application-level
//                confidentiality.
// getBinding -- returns the transaction binding
// getSignedProposal -- returns the SignedProposal object, which contains all
//                      data elements part of a transaction proposal
// getTxTimestamp -- returns the timestamp when the transaction was created
//                   This is extracted from transaction ChannelHeader
// setEvent -- If the transaction is validated and successfully committed,
//             the event will be delivered to the current event listeners

const shim = require('fabric-shim');
const util = require('util');
const path = require('path');

var Chaincode = class {

  // Initialize the chaincode
  async Init(stub) {
    let fileName = __filename.slice(__filename.lastIndexOf(path.sep)+1, __filename.length -3);
    console.info('========= Instantiated chaincode '+fileName+' =========');
    return shim.success();
  }

  // Invoke, to update or query the ledger in a proposal transaction.
  async Invoke(stub) {
    let ret = stub.getFunctionAndParameters();
    let method = this[ret.fcn];
    if (!method) {
      console.log('no method of name:' + ret.fcn + ' found');
      return shim.success();
    }
    try {
      let payload = await method(stub, this, ret.params);
      return shim.success(payload);
    } catch (err) {
      console.log(err);
      return shim.error(err);
    }
  }

  // ========================================================
  // Input Sanitation - input checking, look for empty strings
  // ========================================================
  sanitizeArgs(args, count) {
    if (args.length != count){
      throw new Error('Incorrect number of arguments. Expecting '+count);
    }
  	for (let i=0;i<args.length;i++) {
  		if (!args[i]) {
  			throw new Error(args[i]+ ' must be a non-empty string');
  		}
  	}
  }

  //GetArgs returns the arguments intended for the chaincode Init and Invoke
	// as an array of byte arrays
  async getArgs(stub, thisClass) {
    console.info('\n---------- getArgs ----------');
    let args = stub.getArgs();
    console.log(args);
    thisClass.sanitizeArgs(args, 4);
    let A = args[1];
    let B = args[2];
    let C = args[3];
    return Buffer.from(A +' '+B+' '+C);
  }

  async getStringArgs(stub, thisClass) {
    console.info('\n---------- getStringArgs ----------');
    let args = stub.getStringArgs();
    console.log(args);
    thisClass.sanitizeArgs(args, 4);
    let A = args[1];
    let B = args[2];
    let C = args[3];
    return Buffer.from(A +' '+B+' '+C);
  }

  async getFunctionAndParameters(stub, thisClass) {
    console.info('\n---------- getFunctionAndParameters ----------');
    let ret = stub.getFunctionAndParameters();
    let args = ret.params;
    console.log(args);
    thisClass.sanitizeArgs(args, 3);
    let A = args[0];
    let B = args[1];
    let C = args[2];
    return Buffer.from(ret.fcn + ' '+A +' '+B+' '+C);
  }

  async getTxID(stub) {
    console.info('\n---------- getTxID ----------');
    let txId = stub.getTxID();
    console.info(txId);
    return Buffer.from(txId);
  }

  async invokeChaincode(stub, thisClass, args) {
    console.info('\n---------- invokeChaincode ----------');
    thisClass.sanitizeArgs(args, 3);
    let chaincodeName = args[0];
    let ccArgs = [];
    ccArgs.push(args[1]);
    ccArgs.push(args[2]);
    let channel = '';
    let response = await stub.invokeChaincode(chaincodeName, ccArgs, channel);
    console.log(response);
    if (response.status != 200){
      throw new Error('Failed calling chaincode '+chaincodeName);
    } else {
      console.log(response.payload.buffer.toString());
      //TODO: Parse the response
      return Buffer.from(response.payload.buffer.toString());
    }
  }

  async getCreator(stub) {
    console.info('\n---------- getCreator ----------');
    let creator = stub.getCreator();
    console.log(creator);
    return Buffer.from(creator.mspid);
  }

  async getTransient(stub, thisClass, args) {
    console.info('\n---------- getTransient ----------');
    let transientMap = stub.getTransient();
    console.log(transientMap);
    //TODO: Parse transientMap and send it
    return Buffer.from('');
  }

  //TODO: what do we want to verify for this API in Behave
  async getBinding(stub) {
    console.info('\n---------- getBinding ----------');
    let binding = stub.getBinding();
    console.log(binding);
    return Buffer.from(binding.toString());
  }

  //TODO: what do we want to verify for this API in Behave
  async getSignedProposal(stub){
    console.info('\n---------- getSignedProposal ----------');
    let signedProposal = stub.getSignedProposal();
    console.log(signedProposal);
    //TODO: Parse signedProposal
    return Buffer.from(signedProposal.toString());
  }

  //TODO: FAB-6417 - This has to be fixed to make this work
  async getTxTimestamp(stub){
    console.info('\n---------- getTxTimestamp ----------');
    let timestamp = stub.getTxTimestamp();
    console.log(timestamp);
    return Buffer.from(timestamp);
  }

  //TODO: can test this when there is even handling on the client side
  async setEvent(stub, thisClass, args){
    console.info('\n---------- setEvent ----------');
    thisClass.sanitizeArgs(args, 2);
    let name = args[0];
    let payload = args[1];
    stub.setEvent(name, payload);
  }
};

shim.start(new Chaincode());

File: app/platform/fabric/e2e-test/chaincodes/shimApiDriver/go/shimAPIDriver.go
/*
#
# copyright IBM Corp. All Rights Reserved.
#
# SPDX-License-Identifier: Apache-2.0
#
*/

package main

import (
	"bytes"
	"encoding/gob"
	"fmt"
	"github.com/golang/protobuf/proto"
	"github.com/hyperledger/fabric/common/tools/protolator"
	"github.com/hyperledger/fabric/protos/msp"
	"time"

	"github.com/hyperledger/fabric/core/chaincode/shim"
	pb "github.com/hyperledger/fabric/protos/peer"
)

// SimpleChaincode example simple Chaincode implementation
type SimpleChaincode struct {
}

// ===================================================================================
// Main
// ===================================================================================
func main() {
	err := shim.Start(new(SimpleChaincode))
	if err != nil {
		fmt.Printf("Error starting Simple chaincode: %s", err)
	}
}

// Init initializes chaincode
// ===========================
func (t *SimpleChaincode) Init(stub shim.ChaincodeStubInterface) pb.Response {
	return shim.Success(nil)
}

// Invoke - Our entry point for Invocations
// ========================================
func (t *SimpleChaincode) Invoke(stub shim.ChaincodeStubInterface) pb.Response {

	fmt.Println("\n\nex02 Invoke")

	fmt.Printf("Begin*** GetArgs \n")
	//check for getArgs and getSttringArgs
	args_take1 := stub.GetArgs()

	for key, currArg := range args_take1 {
		fmt.Printf("args_take1[%d] := %s\n", key, string(currArg))
	}
	fmt.Printf("End*** GetArgs \n\n")

	fmt.Printf("Begin*** GetStringArgs \n")
	params := stub.GetStringArgs()
	fmt.Printf("args_take2 := %s \n", params)
	fmt.Printf("End*** GetStringArgs \n\n")

	fmt.Printf("Begin*** GetArgsSlice \n")
	argsSlice, err := stub.GetArgsSlice()
	if err != nil {
		fmt.Printf("Error in argsSlice := %v \n", err)
	}

	if err == nil {
		fmt.Printf("argsSlice := %v \n", string(argsSlice))
	}
	fmt.Printf("End*** GetArgsSlice\n\n")

	fmt.Printf("Begin*** GetFunctionAndParameters \n")
	function, args := stub.GetFunctionAndParameters()
	fmt.Printf("function := %s, args := %s \n", function, args)
	fmt.Printf("End*** GetFunctionAndParameters\n\n")

	// Handle different functions
	if function == "getTxTimeStamp" {
		return t.getTxTimeStamp(stub)
	} else if function == "getCreator" {
		return t.getCreator(stub)
	} else if function == "getBinding" {
		return t.getBinding(stub)
	} else if function == "getSignedProposal" {
		return t.getSignedProposal(stub)
	} else if function == "getTransient" {
		return t.getTransient(stub)
	}

	fmt.Println("invoke did not find func: " + function) //error
	return shim.Error("Received unknown function invocation")
}

//===================================================================================================
// functon getCreator
// You can verify by calling getCreator during initMarble and checking fot the value
// during a transferMarble say
//===================================================================================================

func (t *SimpleChaincode) getCreator(stub shim.ChaincodeStubInterface) pb.Response {

	fmt.Printf("\nBegin*** getCreator \n")
	creator, err := stub.GetCreator()
	if err != nil {
		fmt.Printf("GetCreator Error")
		return shim.Error(err.Error())
	}

	si := &msp.SerializedIdentity{}
	err2 := proto.Unmarshal(creator, si)
	if err2 != nil {
		fmt.Printf("Proto Unmarshal Error")
		return shim.Error(err2.Error())
	}
	buf := &bytes.Buffer{}
	protolator.DeepMarshalJSON(buf, si)
	fmt.Printf("End*** getCreator \n")
	fmt.Printf(string(buf.Bytes()))

	return shim.Success([]byte(buf.Bytes()))
}

//===================================================================================================
// functon to getBinding
//===================================================================================================
func (t *SimpleChaincode) getBinding(stub shim.ChaincodeStubInterface) pb.Response {
	fmt.Printf("\nBegin*** getBinding \n")
	binding, err := stub.GetBinding()
	if err != nil {
		fmt.Printf("Returning error ****************** ")
		return shim.Error(err.Error())
	} else if binding == nil {
		fmt.Printf("###### No Transaction Binding is generated ###### ")
		return shim.Error("###### No Transaction Binding is generated ###### ")
	}
	fmt.Printf("\t returned value from stub : %v\n", binding)
	fmt.Printf("End*** getBinding \n")
	return shim.Success(binding)
}

//===================================================================================================
// functon to getTxTimestamp
// in the time that is associated with current invoke on channel
//===================================================================================================
func (t *SimpleChaincode) getTxTimeStamp(stub shim.ChaincodeStubInterface) pb.Response {
	fmt.Printf("\nBegin*** getTxTimeStamp \n")
	txTimeAsPtr, err := stub.GetTxTimestamp()
	if err != nil {
		fmt.Printf("Returning error ******************\n")
		return shim.Error(err.Error())
	}
	fmt.Printf("\t returned value from stub: %v\n", txTimeAsPtr)
	fmt.Printf("\t After converting time to Unix format %s \n", time.Unix(txTimeAsPtr.Seconds, int64(txTimeAsPtr.Nanos)).String())
	fmt.Printf("\nEnd*** getTxTimeStamp \n")
	//return shim.Success([]byte(txTimeAsPtr))
	return shim.Success(nil)
}

//===================================================================================================
// functon to getTransient
// got to pass these variables during invoke in a transient map
// these values are not stored on ledger
//===================================================================================================
func (t *SimpleChaincode) getTransient(stub shim.ChaincodeStubInterface) pb.Response {
	fmt.Printf("\nBegin*** getTransient \n")
	payload, err := stub.GetTransient()
	fmt.Printf(" payload from chaincode : %v", payload)
	if err != nil {
		return shim.Error(err.Error())
	}
	for key, currArg := range payload {
		fmt.Printf("Inside ... Loop")
		fmt.Printf("payload[%d] := %s\n", key, currArg)
	}
	b, err2 := GetBytes(payload)
	if err2 != nil {
		return shim.Error(err2.Error())
	}
	fmt.Printf("\nEnd*** getTransient \n")
	return shim.Success([]byte(b))
}

//===================================================================================================
// functon to getSignedProposal
//===================================================================================================
func (t *SimpleChaincode) getSignedProposal(stub shim.ChaincodeStubInterface) pb.Response {
	fmt.Printf("\nBegin*** getSignedProposal \n")
	signedProposal, err := stub.GetSignedProposal()
	if err != nil {
		fmt.Printf("Returning error ****************** ")
		return shim.Error(err.Error())
	}
	fmt.Printf("\t returned value from stub: %v", signedProposal)
	fmt.Printf("\nEnd*** getSignedProposal \n")
	buf := &bytes.Buffer{}
	protolator.DeepMarshalJSON(buf, signedProposal)
	fmt.Printf(string(buf.Bytes()))
	return shim.Success([]byte(buf.Bytes()))
}

func GetBytes(key interface{}) ([]byte, error) {
	var buf bytes.Buffer
	enc := gob.NewEncoder(&buf)
	err := enc.Encode(key)
	if err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}
